flows/definition/legacy_test.go

```diff
@@ -127,24 +187,90 @@ func TestTestMigration(t *testing.T) {
 				t.Errorf("Got case:\n%s\n\nwhen expecting:\n%s\n\n", migratedCaseJSON, expectedCaseJSON)
 			}
 
-			checkFlowLocalization(t, &migratedFlow, test.ExpectedLocalization)
+			checkFlowLocalization(t, &migratedFlow, test.ExpectedLocalization, migratedCaseRaw)
+		}
+	}
+}
+
+func TestRuleSetMigration(t *testing.T) {
+	data, err := ioutil.ReadFile("testdata/ruleset_migrations.json")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var tests []RuleSetMigrationTest
+	err = json.Unmarshal(data, &tests)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, test := range tests {
+		legacyFlowsJSON := fmt.Sprintf(legacyRuleSetHolderDef, string(test.LegacyRuleSet))
+		legacyFlows, err := ReadLegacyFlows(json.RawMessage(legacyFlowsJSON))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		migratedFlow := legacyFlows[0]
+
+		// check we now have a new node in addition to the 3 actionsets used as destinations
+		if len(migratedFlow.Nodes()) <= 3 {
+			t.Errorf("Got no migrated nodes from legacy ruleset:\n%s\n\n", string(test.LegacyRuleSet))
+		} else {
+			// find the new node which might be before or after the actionset nodes
+			var migratedNode flows.Node
+			for _, node := range migratedFlow.Nodes() {
+				if node.UUID() != "5b977652-91e3-48be-8e86-7c8094b4aa8f" && node.UUID() != "833fc698-d590-42dc-93e1-39e701b7e8e4" && node.UUID() != "42ff72d3-5f4d-4dbf-89c9-8a97864dabcd" {
+					migratedNode = node
+				}
+			}
+
+			migratedNodeRaw, _ := json.Marshal(migratedNode)
+			migratedNodeJSON := formatJSON(migratedNodeRaw)
+			expectedNodeJSON := formatJSON(test.ExpectedNode)
+
+			if !wildcardEquals(migratedNodeJSON, expectedNodeJSON) {
+				t.Errorf("Got node:\n%s\n\nwhen expecting:\n%s\n\n", migratedNodeJSON, expectedNodeJSON)
+			}
+
+			checkFlowLocalization(t, &migratedFlow, test.ExpectedLocalization, migratedNodeRaw)
 		}
 	}
 }
 
-func checkFlowLocalization(t *testing.T, flow *LegacyFlow, expectedLocalization json.RawMessage) {
+func checkFlowLocalization(t *testing.T, flow *LegacyFlow, expectedLocalizationRaw json.RawMessage, substitutionSource json.RawMessage) {
 	actualLocalization := *flow.translations.(*flowTranslations)
 	actualLocalizationRaw, _ := json.Marshal(actualLocalization)
 	actualLocalizationJSON := formatJSON(actualLocalizationRaw)
-	expectedLocalizationJSON := formatJSON(expectedLocalization)
+
+	// Because localization keys are UUIDs and some of those may be generated during migration, ordering of localized
```

Comment:
I think Eric was talking about adding UUIDs to our current Actions in our flow defs in order to make this a bit simpler and more predictable. (I think this was also needed to simplify merging all the msgs from a broadcast together)